cunumeric/module.py

```diff
@@ -3506,6 +3506,46 @@ def put(
     a.put(indices=indices, values=values, mode=mode)
 
 
+@add_boilerplate("a", "mask", "values")
+def putmask(a: ndarray, mask: ndarray, values: ndarray) -> None:
+    """
+    putmask(a, mask, values)
+    Changes elements of an array based on conditional and input values.
+    Sets ``a.flat[n] = values[n]`` for each n where ``mask.flat[n]==True``.
+    If `values` is not the same size as `a` and `mask` then it will repeat.
+    This gives behavior different from ``a[mask] = values``.
+
+    Parameters
+    ----------
+    a : ndarray
+        Target array.
+    mask : array_like
+        Boolean mask array. It has to be the same shape as `a`.
+    values : array_like
+        Values to put into `a` where `mask` is True. If `values` is smaller
+        than `a` it will be repeated.
+
+    See Also
+    --------
+    numpy.putmask
+
+    Availability
+    ------------
+    Multiple GPUs, Multiple CPUs
+    """
+    if not a.shape == mask.shape:
+        raise ValueError("mask and data must be the same size")
+
+    mask = mask._warn_and_convert(np.dtype(bool))
+
+    if a.dtype != values.dtype:
+        values = values._warn_and_convert(a.dtype)
+    if values.shape != a.shape and values.size != 1:
+        values = values._wrap(a.size)
+        values = values.reshape(a.shape)
```

Comment:
what's the benefit of doing this over tiling followed by slicing? if I understand it correctly, this code creates three temporary stores, one for indices, one for the copy output, and one from the final reshaping. In comparison, the tiling would materialize only one temporary. Plus, tiling is a much simpler operation than the gather copy you do in `_wrap`. Please consider rewriting this with slicing.

Reply:
@magnatelee , consider this example:
```
import numpy as np

arr = np.array([[[1, 2, 3, 4, 5]]])
ind = np.array([[[True,True, False, False,True]]])
v = np.array([[11,12],[13,14]])
np.putmask(arr,ind,v)
print(arr)
```
this will output:
```
[[[11 12  3  4 11]]]
```
In this case shape of v is 2d and shape of mask is 3d. In this case we could do following instead of `_wrap`:
```
v_flat = v.ravel()
old_len = v.size
new_len = a.size
reps = (new_len + old_len - 1) // old_len
v_new = v_flat.tile(reps) if reps > 1 else v_flat
v_new = v_new[:new_len].copy()
v_new = v.reshape(a.shape)
```
This approach would avoid gather/scatter copy, however there is a risk of allocating too much memory for v_new if `new_len -old_len` is a small value and `old_len` is very large

Should I anyway go with this `tile` approach?

Reply:
First of all, I think it's worth adding an optional out or shape parameter to `tile` to express a tiling up to a shape. Then we don't need to worry about `v_new` getting too big. Second, an input array to `tile` being too big is already a problem and we can potentially fix that by falling back to `cunumeric.block. Or we could use `_wrap` in that case. In either case, I feel we can't get away with it and some heuristic based on the size is inevitable in choosing the right implementation.

Reply:
that said, I guess we should address it in a follow-on PR and not this one, so maybe we just keep the current code. I expect the common case to be those where the `values` array can be easily broadcasted to match the target's shape, so we should focus on making them run fast.

Reply:
sounds good!